From de3afeafbc5397711eb1c22367d286bd02dbcaab Mon Sep 17 00:00:00 2001
From: Howard Hsu <howard-yh.hsu@mediatek.com>
Date: Mon, 27 Jul 2026 15:04:31 +0000
Subject: [PATCH] wifi: mt76: mt7996: reject iTWT setup requests from MLD
 stations

The firmware does not support individual TWT agreements with non-AP MLDs,
and the driver only tracks TWT flow state on the default link, which may
not be the link the agreement was negotiated on. Reject TWT setup
requests from MLD stations instead of programming an unsupported
configuration.

Signed-off-by: Howard Hsu <howard-yh.hsu@mediatek.com>
Link: https://patch.msgid.link/20260727150434.1778520-12-nbd@nbd.name
Signed-off-by: Felix Fietkau <nbd@nbd.name>
---
 mt7996/mac.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/mt7996/mac.c b/mt7996/mac.c
index 4794dacb3756..1cba381a78b8 100644
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -3218,6 +3218,12 @@ void mt7996_mac_add_twt_setup(struct ieee80211_hw *hw,
 	struct mt7996_twt_flow *flow;
 	u8 flowid, table_id, exp;
 
+	/* the firmware does not support iTWT agreements with MLD peers, and
+	 * driver TWT state is only tracked on the default link
+	 */
+	if (sta->mlo)
+		goto out;
+
 	if (mt7996_mac_check_twt_req(twt))
 		goto out;
 
-- 
2.43.0

